test/cpp/api/sequential.cpp

```diff
@@ -86,40 +86,40 @@ TEST_CASE("sequential") {
 
     Sequential sequential;
     for (auto& module : modules) {
-      sequential.push_back(module);
+      sequential->push_back(module);
     }
-    REQUIRE(sequential.size() == 3);
+    REQUIRE(sequential->size() == 3);
 
     SECTION("at()") {
       SECTION("returns the correct module for a given index") {
         for (size_t i = 0; i < modules.size(); ++i) {
-          REQUIRE(&sequential.at<M>(i) == modules[i].get());
+          REQUIRE(&sequential->at<M>(i) == modules[i].get());
         }
       }
       SECTION("throws for a bad index") {
         REQUIRE_THROWS_WITH(
-            sequential.at<M>(modules.size() + 1),
+            sequential->at<M>(modules.size() + 1),
             StartsWith("Index out of range"));
         REQUIRE_THROWS_WITH(
-            sequential.at<M>(modules.size() + 1000000),
+            sequential->at<M>(modules.size() + 1000000),
             StartsWith("Index out of range"));
       }
     }
 
     SECTION("ptr()") {
       SECTION("returns the correct module for a given index") {
         for (size_t i = 0; i < modules.size(); ++i) {
-          REQUIRE(sequential.ptr(i).get() == modules[i].get());
-          REQUIRE(sequential[i].get() == modules[i].get());
-          REQUIRE(sequential.ptr<M>(i).get() == modules[i].get());
+          REQUIRE(sequential->ptr(i).get() == modules[i].get());
+          REQUIRE((*sequential)[i].get() == modules[i].get());
```

Comment:
Making the modules use `->` everywhere is definitely a good step towards improving usability, but this one is a step back. Couldn't we implement `operator[]` in the holder that would redirect the call to the module?

Reply:
I tried this, but struggled to define it in a way such that the return type is the return type of the delegated call. I can try again, maybe I figure it out. certainly `-> decltype((*impl_)[args...])` did not work

Reply:
Ok I fixed it. I had to move `impl_` to the top of the class, otherwise I can't use it in the trailing return type of the methods. TIL!

Reply:
C++ is a beautiful language :)